lang/wasm.cc

```diff
@@ -290,51 +195,60 @@ struct from_val_visitor {
+        create_func(*inst, *store, ctx.function_name);
+    } catch (const rust::Error& e) {
+        throw wasm::exception(e.what());
+    }
+}
 seastar::future<bytes_opt> run_script(context& ctx, wasmtime::Store& store, wasmtime::Instance& instance, wasmtime::Func& func, const std::vector<data_type>& arg_types, const std::vector<bytes_opt>& params, data_type return_type, bool allow_null_input) {
     wasm_logger.debug("Running function {}", ctx.function_name);
 
-    // Replenish the store with initial amount of fuel
-    auto added = store.context().add_fuel(ctx.engine_ptr->initial_fuel_amount());
-    if (!added) {
-        co_await coroutine::return_exception(wasm::exception(added.err().message()));
-    }
-    std::vector<wasmtime::Val> argv;
+    rust::Box<wasmtime::ValVec> argv = wasmtime::get_val_vec();
     for (size_t i = 0; i < arg_types.size(); ++i) {
         const abstract_type& type = *arg_types[i];
         const bytes_opt& param = params[i];
         // If nulls are allowed, each type will be passed indirectly
         // as a struct {bool is_null; int32_t serialized_size, char[] serialized_buf}
         if (allow_null_input) {
-            visit(type, init_nullable_arg_visitor{param, argv, store, instance});
+            visit(type, init_nullable_arg_visitor{param, *argv, store, instance});
         } else if (param) {
-            visit(type, init_arg_visitor{param, argv, store, instance});
+            visit(type, init_arg_visitor{param, *argv, store, instance});
         } else {
             co_await coroutine::return_exception(wasm::exception(format("Function {} cannot be called on null values", ctx.function_name)));
         }
     }
-    uint64_t fuel_before = *store.context().fuel_consumed();
-
-    auto result = func.call(store, argv);
-
-    uint64_t consumed = *store.context().fuel_consumed() - fuel_before;
-    wasm_logger.debug("Consumed {} fuel units", consumed);
-
-    if (!result) {
-        co_await coroutine::return_exception(wasm::instance_corrupting_exception("Calling wasm function failed: " + result.err().message()));
-    }
-    std::vector<wasmtime::Val> result_vec = std::move(result).unwrap();
-    if (result_vec.size() != 1) {
-      co_await coroutine::return_exception(wasm::exception(format("Unexpected number of returned values: {} (expected: 1)", result_vec.size())));
+    auto rets = wasmtime::get_val_vec();
+    rets->push_i32(0);
+
+    auto fut = wasmtime::get_func_future(store, func, *argv, *rets);
+    bool stop = false;
+    while (!stop) {
+        std::exception_ptr eptr;
+        try {
+            stop = fut->resume();
+        } catch (const rust::Error& e) {
+            eptr = std::make_exception_ptr(wasm::instance_corrupting_exception(format("Calling wasm function failed: {}", e.what())));
+        }
+        if (eptr) {
+            co_await coroutine::return_exception_ptr(eptr);
+        }
+        co_await coroutine::maybe_yield();
```

Comment:
:relieved:

Reply:
I see you do yield here. Why are the others different? Need comments.

Reply:
I don't yield in other places mostly because they should finish quickly, the other cases are also not used in future functions so it would require quite a bit more changes to yield there - I'll add comments.